### PR TITLE
add letter_contact_block edit fields

### PR DESCRIPTION
**Pull request by leohemsted**
no actual template functionality yet - just the ability for services that have letters enabled to edit a 10 line block that will in the future go on the top right hand side of their letters with contact information

text on the edit page probably needs a bit of TLC tho

image

### CODE REVIEW

**Comment by quis** [comment]
# Make letter contact textbox half-wide

This approximates the width of the contact block area in the letter template itself.

image

File: app/main/views/service_settings.py [hunk 1]
```diff
@@ -259,6 +261,28 @@ def service_set_sms_sender(service_id):
         form=form)
 
 
+@main.route("/services/<service_id>/service-settings/set-letter-contact-block", methods=['GET', 'POST'])
+@login_required
+@user_has_permissions('manage_settings', admin_override=True)
+def service_set_letter_contact_block(service_id):
+    print(current_service)
+    print(current_service['can_send_letters'])
```
Comment (imdadahad) [inline app/main/views/service_settings.py]: 🖨

File: app/main/forms.py [hunk 2]
```diff
@@ -486,10 +486,20 @@ class ServiceSmsSender(Form):
...
 
     def validate_sms_sender(form, field):
-        if field.data and not re.match('^[a-zA-Z0-9\s]+$', field.data):
+        if field.data and not re.match(r'^[a-zA-Z0-9\s]+$', field.data):
             raise ValidationError('Use letters and numbers only')
 
 
+class ServiceLetterContactBlock(Form):
+    letter_contact_block = TextAreaField(
+        'Letter contact block'
```
Comment (imdadahad) [inline app/main/forms.py]: Does this need a RequiredValidator?

Comment (leohemsted) [inline app/main/forms.py]: it's not required - if you leave it empty, that's okay

File: app/main/views/service_settings.py [hunk 3]
```diff
@@ -259,6 +261,28 @@ def service_set_sms_sender(service_id):
         form=form)
 
 
+@main.route("/services/<service_id>/service-settings/set-letter-contact-block", methods=['GET', 'POST'])
```
Comment (imdadahad) [inline app/main/views/service_settings.py]: Slightly minor, but why is this endpoint prefixed with set? I know we do this with other endpoints but doesn't seem right given it's a `GET` too.

File: app/main/forms.py (hunk 2 above)
Comment (imdadahad) [inline app/main/forms.py]: 👍

File: tests/app/main/views/test_service_settings.py [hunk 4]
```diff
@@ -538,6 +580,61 @@ def test_if_sms_sender_set_then_form_populated(
     assert page.find(id='sms_sender')['value'] == 'elevenchars'
 
 
+@pytest.mark.parametrize('method', ['get', 'post'])
+def test_cant_set_letter_contact_block_if_service_cant_send_letters(
```
Comment (imdadahad) [inline tests/app/main/views/test_service_settings.py]: Being pedantic here but `get` -> `GET` and with post?

File: app/main/views/service_settings.py (hunk 3 above)
Comment (leohemsted) [inline app/main/views/service_settings.py]: well, i guess the pages' purpose is to set the value of the block - it has a GET just cos the page has to be loaded and that's the WTForms pattern.

if you want to view the services' letter contact block, you'd just look at it on the service-settings page

or something idk /shrug

File: tests/app/main/views/test_service_settings.py (hunk 4 above)
Comment (leohemsted) [inline tests/app/main/views/test_service_settings.py]: ah those are the actual functions, eg `logged_in_client.get(...)`. `logged_in_client.GET` will error

Comment (imdadahad) [inline tests/app/main/views/test_service_settings.py]: ah ofc, missed the `getattr`

File: app/main/views/service_settings.py (hunk 3 above)
Comment (imdadahad) [inline app/main/views/service_settings.py]: fair enough

**Comment by imdadahad** [comment]
Happy to approve once those prints are removed 👍

**Comment by leohemsted** [comment]
pending chris' work ofc

**Comment by quis** [comment]
@imdadahad Can you review the commits I just added? 68113e7, 63f5239, 3dabd39, 92a51c3

Looks like this:
image

**Comment by imdadahad** [comment]
@quis looks good 👍

**Pull request merged**